### ISSUE TITLE
findOne gives type <T> but can be null

### ISSUE THREAD

**Issue by Mudbill** [bajankristof/nedb-promises#55]
### Problem:
Basically, the `findOne` call assumes something is found whether or not it actually was, and TypeScript therefore doesn't alert you of trying to get properties from null.

```js
const data = await db.findOne<{ name: string }>({ id: 1 });
console.log(data.name); // No TypeScript warning, but throws a TypeError if no row has id = 1
```
The returned type becomes `<T & Document>` but I think it should be `<(T & Document) | null>`. Alternatively throw some error.

```js
const data = await db.findOne<{ name: string }>({ id: 1 });
console.log(data.name); // TypeScript warning: Object is possibly 'null'
if (data) {
  console.log(data.name) // Safe
}
```

This may also apply to some other similar methods.

Side note: Giving it a type like `findOne<T | null>()` doesn't work either, it still becomes `T & Document`

### Possible solution:
Change the return type for `findOne` from `Promise<T & Document>` to `Promise<(T & Document) | null>`. This provides the desired effect, but I haven't delved into possible side effects and may be considered a breaking change for users who don't do this type check in their code already (although they are at risk of TypeErrors if not)

**Comment by bajankristof** [comment]
Hi,
Thanks for opening this issue, you're absolutely right about this. One question, did you test the modification (`(T & Document) | null`) in your project, does this actually solve the problem?

**Comment by Mudbill** [comment]
Yes, I modified the file locally in `node_modules` and it worked. I then instead put it in `@types/nedb-promises.d.ts`. I actually have a [repo with this file present](https://github.com/Mudbill/ts-rest-api/blob/master/src/%40types/nedb-promises.d.ts), though as mentioned there may be a few other functions that could use a similar edit.

**Comment by bajankristof** [comment]
Alright, I got a bit of time, I'm taking a look at the TS definition file to see if there's anything that should be changed.

**Comment by bajankristof** [comment]
Okay, so there's actually a lot of things that are outdated and/or could be better in the TypeScript definition file so this will definetly be a new major release. I will update this issue once I'm done with it.

**Comment by bajankristof** [comment]
The TypeScript declaration file just got a major overhaul in `6.0.0` which is now published to NPM.

You can see the changes here: https://github.com/bajankristof/nedb-promises/commit/2abb970f01d52ff8ce0af204666f1c8b4d52432a

I tested the autocompletion, it is much smarter now than it was before regarding document fields, projections, updates, timestamps, etc.

I also added a missing indexing option in the declaration file (`expireAfterSeconds`).

Please close the issue if you found the problem solved on your end too.

Thanks!

**Comment by bajankristof** [comment]
(In the meantime I also published `6.0.1` with an updated class description.)

**Comment by Mudbill** [comment]
Nice work! At a glance it looks to resolve the original issue – the resulting type is now `({ _id: string } & T) | null` which works great. Thanks for the quick responses :+1:
